README.md
# MyApp::Weather

A [Mojolicious](https://mojolicious.org/) web application that gets weather
forecasts from [MET Norway](https://api.met.no/) and produces web calendars in
the iCalendar data format.  Calendars can be subscribed from software such as
[Thunderbird](https://www.thunderbird.net/) and
[ICSx⁵](https://icsx5.bitfire.at/).

    env WEATHER_USER_AGENT='example.com support@example.com' \
        script/weather daemon -l http://localhost:3000
    curl http://localhost:3000/Oslo.ics

Calendar events have got a summary consisting of emojis and temperatures as
well as a longer description of the weather conditions.

    BEGIN:VEVENT
    UID:58d61e67800ce518ca9eef11d0347fa3e6ea3a68
    DTSTAMP:20211106T020745Z
    DTSTART;VALUE=DATE:20211107
    SUMMARY:🌫☀️ 6/-1°
    LOCATION:Oslo
    DESCRIPTION:00-05:   2° Fair\n05-10:   1° Fog\n10-12:   4° Partly clo
     udy\n12-18:   6° Clear sky\n18-22:   2° Clear sky\n22-24:  -1° Fair
    COMMENT:Based on data from MET Norway
    CLASS:PUBLIC
    TRANSP:TRANSPARENT
    X-SOGO-SEND-APPOINTMENT-NOTIFICATIONS:NO
    END:VEVENT

## DEPENDENCIES

Requires Perl 5.20, Mojolicious 9, Role::Tiny and
Geo::Location::TimeZoneFinder.

## QUERY PARAMETERS

* lang

    * de (German)
    * en (English)
    * nb (Norwegian)

* temperature_unit

    * C (Celsius)
    * F (Fahrenheit)

Example:

    curl 'http://localhost:3000/Oslo.ics?lang=en&temperature_unit=F'

## CONFIGURATION

### Environment Variables

#### CACHE_DIRECTORY

Weather forecasts and locations are cached in this directory.  The directory
must exist and be writable.

#### LANG

The default weather forecast language.

#### REQUEST_BASE

The base path in the frontend proxy, e.g. /weather.  Empty by default.

#### TEMPERATURE_UNIT

The default temperature unit.

#### TIME_ZONE_DATABASE

The file base of a time zone database from
https://github.com/evansiroky/timezone-boundary-builder, e.g.
/usr/local/share/timezones/combined-shapefile.

#### WEATHER_USER_AGENT

An identifying user agent string.  See https://api.met.no/doc/TermsOfService
and https://operations.osmfoundation.org/policies/nominatim/ for more
information.

## LICENSE AND COPYRIGHT

Copyright (C) 2023 Andreas Vögele

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU Affero General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any
later version.

[MET Norway](https://www.met.no/) provides data under the [Norwegian Licence
for Open Government Data (NLOD) 2.0](https://data.norge.no/nlod/en/2.0/) and
[Creative Commons 4.0 BY International](http://creativecommons.org/licenses/by/4.0)
licenses.


Changes
Revision history for MyApp::Weather

0.006  2023-01-11 10:58:44 CET

  - Use the location's local time
  - Simplify the TimeStep class
  - Use cache directories instead of single files

0.005  2022-09-30 10:08:19 CEST

  - Round the temperatures with int()

0.004  2022-08-17 10:15:15 CEST

  - Fix the temperature rounding
  - Add .github/workflows/linux.yml

0.003  2022-01-07 21:53:46 CET

  - Update dist.ini
  - Bump the copyright year

0.002  2021-11-21 06:53:59 CET

  - Prepend rainy days with a snowflake if there is snow

0.001  2021-09-21 13:47:11 CEST

  - First version



lib/MyApp/Weather.pm
package MyApp::Weather;
use Mojo::Base 'Mojolicious', -signatures;

# SPDX-License-Identifier: AGPL-3.0-or-later

our $VERSION = '0.006';

use Mojo::File qw(curfile);
use Mojo::Home;

has default_location => 'Oslo';

sub startup ($app) {
    # Switch to installable directories.
    $app->home(Mojo::Home->new(curfile->sibling('Weather')));
    $app->static->paths->[0]   = $app->home->child('public');
    $app->renderer->paths->[0] = $app->home->child('templates');

    # Use random secrets since this application doesn't use session cookies.
    $app->secrets([rand]);

    # Map the file extension "ics" to the MIME type "text/calendar".
    $app->types->type(ics => 'text/calendar');

    # Register helpers and hooks.
    $app->plugin('MyApp::Weather::Plugin::Helpers');
    $app->hook(before_dispatch => \&_before_dispatch);

    my $r = $app->routes;

    # Route requests like "/Oslo.ics" to MyApp::Weather::Controller::Location.
    $r->get('/:location' => [format => [qw(html ics)]])->to(
        'Location#calendar',
        format   => 'ics',
        location => $app->default_location,
    );

    return;
}

sub _before_dispatch ($c) {
    # Set the base path in the frontend proxy.
    if (state $base = $ENV{REQUEST_BASE}) {
        my $url = Mojo::URL->new($base);
        if ($url->host) {
            $c->req->url->base($url);
        }
        else {
            $c->req->url->base->path($url->path);
        }
    }

    return;
}

1;
__END__

=encoding UTF-8

=head1 NAME

MyApp::Weather - Weather forecasts in the iCalendar data format

=head1 VERSION

version 0.006

=head1 SYNOPSIS

  env WEATHER_USER_AGENT='example.com support@example.com' \
      script/weather daemon -l http://localhost:3000
  curl http://localhost:3000/Oslo.ics

=head1 DESCRIPTION

A L<Mojolicious> web application that gets weather forecasts from L<MET
Norway|https://api.met.no/> and produces web calendars in the iCalendar data
format.

=head1 SUBROUTINES/METHODS

=head2 startup

  $app->startup;

Sets up the application and the routes.

Requests are routed to L<MyApp::Weather::Controller::Location>.

=head1 DIAGNOSTICS

None.

=head1 CONFIGURATION AND ENVIRONMENT

=head2 Environment variables

=head3 LANG

The default weather forecast language.  See L<MyApp::Weather::View::Messages>
for supported languages.

=head3 CACHE_DIRECTORY

Weather forecasts and locations are cached in this directory.  The directory
must exist and be writable.

=head3 REQUEST_BASE

The base path in the frontend proxy, e.g. /weather.  Empty by default.

=head3 TEMPERATURE_UNIT

The default temperature unit.  Valid values are "C" for Celsius and "F" for
Fahrenheit.

=head3 TIME_ZONE_DATABASE

The file base of a time zone database from
L<https://github.com/evansiroky/timezone-boundary-builder>, e.g.
F</usr/local/share/timezones/combined-shapefile>.

=head3 WEATHER_USER_AGENT

An identifying user agent string.  See
L<https://api.met.no/doc/TermsOfService> and
L<https://operations.osmfoundation.org/policies/nominatim/> for more
information.

=head1 DEPENDENCIES

Requires L<Mojolicious>.

=head1 INCOMPATIBILITIES

None.

=head1 BUGS AND LIMITATIONS

None known.

=head1 AUTHOR

Andreas Vögele E<lt>andreas@andreasvoegele.comE<gt>

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2023 Andreas Vögele

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU Affero General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option) any
later version.

=cut
